### PR TITLE
Add `ethereum-bsc`

### PR DESCRIPTION
**Pull request by AurevoirXavier**
cc @songtianyi

### CODE REVIEW

File: bin/node/runtime/pangolin/src/pallets/bridge_ethereum_bsc.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// --- substrate ---
+use sp_core::U256;
+// --- darwinia ---
+use crate::*;
+use darwinia_bridge_ethereum_bsc::{BSCConfiguration, Config};
+
+frame_support::parameter_types! {
+	pub Configuration: BSCConfiguration = BSCConfiguration {
```
Comment (AurevoirXavier) [inline bin/node/runtime/pangolin/src/pallets/bridge_ethereum_bsc.rs:8]: What cfg should be used here?

File: bin/node/cli/src/chain_spec.rs [hunk 2]
```diff
@@ -550,5 +551,6 @@ fn pangolin_development_genesis() -> pangolin_runtime::GenesisConfig {
 		},
 		darwinia_evm: pangolin_runtime::EVMConfig { accounts: evm_accounts },
 		dvm_ethereum: Default::default(),
+		darwinia_bridge_ethereum_bsc: Default::default(),
```
Comment (AurevoirXavier) [inline bin/node/cli/src/chain_spec.rs:554]: What genesis header should be used here?

File: bin/node/cli/src/chain_spec.rs [hunk 3]
```diff
@@ -393,6 +393,7 @@ fn pangolin_build_spec_genesis() -> pangolin_runtime::GenesisConfig {
...
```
Comment (AurevoirXavier) [inline bin/node/cli/src/chain_spec.rs:396]: What genesis header should be used here?

File: bin/node/cli/src/chain_spec.rs (hunk 2 above)
Comment (songtianyi) [reply bin/node/cli/src/chain_spec.rs:554]: lastest finalized checkpoint header. relayer and this pallet should start with the same header number

File: bin/node/runtime/pangolin/src/pallets/bridge_ethereum_bsc.rs (hunk 1 above)
Comment (songtianyi) [reply bin/node/runtime/pangolin/src/pallets/bridge_ethereum_bsc.rs:8]: https://github.com/darwinia-network/parity-bridges-common/blob/3fa0052c2d63b938b8c2a18283f206f0e47cdd83/modules/ethereum-bsc/src/mock.rs#L101

example here.
These parameters are the same as the prod environment I think, more like constants

File: Cargo.lock [hunk 4]
```diff
@@ -1441,6 +1463,23 @@ dependencies = [
  "sp-runtime",
 ]
 
+[[package]]
+name = "darwinia-bridge-ethereum-bsc"
```
Comment (HackFisher) [inline Cargo.lock:1467]:
```suggestion
name = "darwinia-bridge-bsc"
```

File: bin/node/cli/src/chain_spec.rs (hunk 3 above)
Comment (HackFisher) [reply bin/node/cli/src/chain_spec.rs:396]: https://bscscan.com/

For Pangolin, maybe we should use its testnet: https://testnet.bscscan.com/

@songtianyi Please select and use a recent epoc/header as the genesis data.

Comment (HackFisher) [inline bin/node/cli/src/chain_spec.rs:396]:
```suggestion
		darwinia_bridge_bsc: Default::default(),
```

File: frame/bridge/ethereum-bsc/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,58 @@
+[package]
+authors     = ["Darwinia Network <hello@darwinia.network>"]
+description = "A Substrate Runtime module that is able to verify bsc headers and authority set finality."
+edition     = "2018"
+homepage    = "https://darwinia.network/"
+license     = "GPL-3.0"
+name        = "darwinia-bridge-ethereum-bsc"
```
Comment (HackFisher) [inline frame/bridge/ethereum-bsc/Cargo.toml:7]:
```suggestion
name        = "darwinia-bridge-bsc"
```

File: frame/bridge/ethereum-bsc/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Verify unsigned relayed headers and finalize authority set
+		#[pallet::weight(0)]
+		pub fn verify_and_update_authority_set_unsigned(
+			origin: OriginFor<T>,
+			headers: Vec<BSCHeader>,
+		) -> DispatchResultWithPostInfo {
+			// TODO: should we ignore all checks while number equals 0?
```
Comment (HackFisher) [inline frame/bridge/ethereum-bsc/src/lib.rs:156]: FinalizedCheckpoint should have already been set, and we have checks later for block number. 

We can ignore this.

File: frame/bridge/ethereum-bsc/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+			let mut unsigned_header = header.to_owned();
+
+			unsigned_header.extra_data = signed_data_slice.to_vec();
+
+			let msg = unsigned_header.compute_hash_with_chain_id(chain_id);
+			let pubkey = crypto::secp256k1_ecdsa_recover(&signature, msg.as_fixed_bytes())
+				.map_err(|_| <Error<T>>::RecoverPubkeyFail)?;
+			let creator = bsc_primitives::public_to_address(&pubkey);
+
+			<CreatorCache<T>>::insert(header.compute_hash(), creator);
```
Comment (HackFisher) [inline frame/bridge/ethereum-bsc/src/lib.rs:475]: This does not make sense.

`CreatorCache` is crypto recover local cache, not global, should not be stored on chain in storage.

**Review by HackFisher — CHANGES_REQUESTED** [review]

**Comment by AurevoirXavier** [comment]
@AsceticBear DVM tests failed

**Comment by AsceticBear** [comment]
> @AsceticBear DVM tests failed

Looks like a network problem, the basic RPC can not fetch result. I re-run it and see what happens.

**Comment by AurevoirXavier** [comment]
> > @AsceticBear DVM tests failed
> 
> Looks like a network problem, the basic RPC can not fetch result. I re-run it and see what happens.

I've re-run it many times. Seems something wrong with the `spec_version`

**Comment by AsceticBear** [comment]
I encountered the following error when I tried to run ci test locally:

```sh
2021-06-08 13:15:55 Running in --dev mode, RPC CORS has been disabled.    
2021-06-08 13:15:55 Darwinia Runtime Module Library    
2021-06-08 13:15:55 ✌️  version 2.4.0-e450d16ae-x86_64-linux-gnu    
2021-06-08 13:15:55 ❤️  by Darwinia Network <hello@darwinia.network>, 2018-2021    
2021-06-08 13:15:55 📋 Chain specification: Pangolin    
2021-06-08 13:15:55 🏷 Node name: cowardly-lettuce-1176    
2021-06-08 13:15:55 👤 Role: AUTHORITY    
2021-06-08 13:15:55 💾 Database: RocksDb at /tmp/substrateQbupbh/chains/pangolin/db    
2021-06-08 13:15:55 ⛓  Native runtime: Pangolin-2403 (Pangolin-1.tx1.au1)    
2021-06-08 13:15:55 `pallet_timestamp::UnixTime::now` is called at genesis, invalid value returned: 0    
2021-06-08 13:15:55 [#0] 🗳  Finalized election round with compute ElectionCompute::OnChain.    
2021-06-08 13:15:55 [0] 💸 new validator set of size 1 has been processed for era 1    
2021-06-08 13:15:55 [darwinia-claims] Genesis Claims List is Set to EMPTY    

====================

Version: 2.4.0-e450d16ae-x86_64-linux-gnu

   0: sp_panic_handler::set::{{closure}}
   1: std::panicking::rust_panic_with_hook
             at /rustc/e37a13cc3594004663738bd18d8100e6db9666cf/library/std/src/panicking.rs:595:17
   2: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/e37a13cc3594004663738bd18d8100e6db9666cf/library/std/src/panicking.rs:497:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/e37a13cc3594004663738bd18d8100e6db9666cf/library/std/src/sys_common/backtrace.rs:141:18
   4: rust_begin_unwind
             at /rustc/e37a13cc3594004663738bd18d8100e6db9666cf/library/std/src/panicking.rs:493:5
   5: core::panicking::panic_fmt
             at /rustc/e37a13cc3594004663738bd18d8100e6db9666cf/library/core/src/panicking.rs:92:14
   6: core::option::expect_none_failed
             at /rustc/e37a13cc3594004663738bd18d8100e6db9666cf/library/core/src/option.rs:1300:5
   7: <darwinia_bridge_bsc::pallet::GenesisConfig as frame_support::traits::hooks::GenesisBuild<T>>::build
   8: std::thread::local::LocalKey<T>::with
   9: sp_state_machine::basic::BasicExternalities::execute_with_storage
  10: <pangolin_runtime::GenesisConfig as sp_runtime::BuildStorage>::assimilate_storage
  11: sp_runtime::BuildStorage::build_storage
  12: <sc_chain_spec::chain_spec::ChainSpec<G,E> as sp_runtime::BuildStorage>::build_storage
  13: sc_service::builder::new_full_parts
  14: drml::service::new_partial
  15: drml::service::new_full
  16: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  17: std::thread::local::LocalKey<T>::with
  18: tokio::runtime::thread_pool::ThreadPool::block_on
  19: tokio::runtime::context::enter
  20: sc_cli::runner::Runner<C>::run_node_until_exit
  21: drml::command::run
  22: drml::main
  23: std::sys_common::backtrace::__rust_begin_short_backtrace
  24: std::rt::lang_start::{{closure}}
  25: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/e37a13cc3594004663738bd18d8100e6db9666cf/library/core/src/ops/function.rs:259:13
      std::panicking::try::do_call
             at /rustc/e37a13cc3594004663738bd18d8100e6db9666cf/library/std/src/panicking.rs:379:40
      std::panicking::try
             at /rustc/e37a13cc3594004663738bd18d8100e6db9666cf/library/std/src/panicking.rs:343:19
      std::panic::catch_unwind
             at /rustc/e37a13cc3594004663738bd18d8100e6db9666cf/library/std/src/panic.rs:431:14
      std::rt::lang_start_internal
             at /rustc/e37a13cc3594004663738bd18d8100e6db9666cf/library/std/src/rt.rs:51:25
  26: main
  27: __libc_start_main
  28: _start


Thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Module { index: 46, error: 19, message: Some("CheckpointNoSigner") }', /home/bear/working/darwinia/darwinia-common/frame/bridge/ethereum-bsc/src/lib.rs:137

This is a bug. Please report it at:

	https://github.com/darwinia-network/darwinia-common/issues/new
```

This is the reason for red ci.

**Comment by AurevoirXavier** [comment]
Okay. I know.

File: frame/bridge/ethereum-bsc/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,538 @@
...
+				ensure!(!recently.contains(&signer), <Error::<T>>::SignedRecently);
+
+				recently.insert(signer);
+
+				// enough proof to finalize new authority set
+				if recently.len() >= last_authority_set.len() / 2 {
+					// already have N/2 valid headers signed by different authority separately
+					// finalize new authroity set
+					<FinalizedAuthority<T>>::put(new_authority_set);
+					<FinalizedCheckpoint<T>>::put(checkpoint);
```
Comment (HackFisher) [inline frame/bridge/ethereum-bsc/src/lib.rs:222]: Kind of duplicate codes with the relative part in `verify_and_update_authority_set_signed`, better to extract to an internal call and reuse

**Review by HackFisher — CHANGES_REQUESTED** [review]

**Comment by AurevoirXavier** [comment]
And I think we need a pre-verify process for the unsigned call.

**Comment by HackFisher** [comment]
> And I think we need a pre-verify process for the unsigned call.

Yes if you mean `validate_unsigned`, do you plan to do it here or in another PR?

**Review by HackFisher — APPROVED** [review]

**Comment by AurevoirXavier** [comment]
> > And I think we need a pre-verify process for the unsigned call.
> 
> Yes if you mean `validate_unsigned`, do you plan to do it here or in another PR?

Another PR would be better.

**Comment by HackFisher** [comment]
@songtianyi @AsceticBear Please help review this pull request

**Review by songtianyi — APPROVED** [review]

**Pull request merged**